### PR TITLE
UDQ field and group variables

### PR DESCRIPTION
**Pull request by joakim-hove**
With this PR the code to evaluate UDQ values has been extended/refactored in
many ways:

 - There is an increased awarenwss of variable type, i.e. a well related UDQ
   versus a field related UDQ. The variable type flows through the AST, and it
   is verified that the expression to evaluate and the target variable are of
   compatible types.

 - Have added support for UDQ field variables and UDQ group variables.

 - The UDQSet type and the UDQDefine::eval() have been refactored to multiplex
   on all the UDQ types, and the specialized UDQWellSet has been removed.

With this we support UDQ variables for field, well and group - I assume that will be Good Enough&trade; in the medium range.

### CODE REVIEW

**Comment by joakim-hove** [comment]
Look ma - now; instead of writing `FOPR` in the summary section, you can write:
```
SUMMARY

FUOPR

SCHEDULE 

UDQ
    DEFINE FUOPR SUM(WOPR) /
    UNIT FUOPR 'SM3/DAY' /
/ 
```
and have the `UDQ` virtual machine do the calculation for you - how cool is that!

File: tests/parser/UDQTests.cpp [hunk 1]
```diff
@@ -112,6 +112,7 @@ BOOST_AUTO_TEST_CASE(UDQWellSetTest) {
     for (std::size_t i=0; i < wells.size(); i++)
         BOOST_CHECK_EQUAL(ws3[wells[i]].value(), i*1.0);
 }
+<<<<<<< 337d73fb20ad410d1d0baa28d1cd452ac4f7e60e
```
Comment (akva2) [inline tests/parser/UDQTests.cpp]: hmmmm

Comment (joakim-hove) [reply tests/parser/UDQTests.cpp]: :astonished:  - works on my machine!

Comment (akva2) [reply tests/parser/UDQTests.cpp]: yes, it's removed in 334b985. drop 62bd2d6 and deal with the "conflict" in 334b985

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by joakim-hove** [comment]
jenkins build this please

**Comment by bska** [comment]
Do you expect any more work here (or related areas) before this PR is ready for final review?

**Comment by joakim-hove** [comment]
No - this can be merged, though there will be more UDQ work coming.

**Comment by joakim-hove** [comment]
> Why use fnmatch() instead of a more regular string matching algorithm?

The `fnmatch()` directly implements/supports the well wildcard matching in eclipse well names.

**Comment by bska** [comment]
> > Why use fnmatch() instead of a more regular string matching algorithm?
> 
> The `fnmatch()` directly implements/supports the well wildcard matching in eclipse well names.

Right.  I now see that we're using that function elsewhere too.  Ideally I think we shouldn't be using POSIX functions directly, but I'm not going to request a different implementation.

**Pull request merged**